### ISSUE TITLE
NBA Jam SRAM getting overwritten at title screen.

### ISSUE THREAD

**Issue by birdybro** [MiSTer-devel/Genesis_MiSTer#163]
From - https://github.com/MiSTer-devel/Main_MiSTer/issues/323

Moving to here since it's not related to main.

Just tested this with NBA Jam (USA, Europe) Rev1 no-intro matching with ROM Storage set to Auto due to the other issues noted with ROM storage switch and SDRAM issues currently on Genesis core. I noticed the same behavior either way, it didn't keep the records. The save file upon load of the game does get updated, so I wonder if it's getting overwritten as if it's a new save. 

Launched with DDR3, 

I also tested for when it SRAM writes, kept opening OSD which will check changes and if changes it will write to .sav file, but if no changes it does nothing. Kept refreshing for date modified in winscp while opening and closing OSD a bunch... and saw changes once the game gets to the title screen. So NBA Jam Rev 1 writes to the SRAM upon getting to the title screen, and the current method implemented in the MiSTer makes this wipe out the records saved in the game.

To test whether or not the genesis core is reading the accurately saved RAM file (or if the save is just not saving any data correctly, or something like that), I then loaded a game, generated the new record in the game, opened OSD to save the record, and did a cold reboot, went back into the genesis core, loaded the game and didn't load the OSD after that until I opened records. This resulted in  

I tested OSD auto save while in a match, it still has no new data to the sram. A match must be concluded in order to update the SRAM. The update happens when "Final Score" is shown.

Both these games used different SRAM techniques. NBA Jam TE was supposed to be accounted for here - https://github.com/MiSTer-devel/Genesis_MiSTer/blob/6cdd10376ec6c4a88d55d1571685307dfd908bd8/Genesis.sv#L853 .  

Original NBA Jam cart, with an already supported SRAM without battery method:

![image](https://user-images.githubusercontent.com/16388068/104044595-bf3a1400-519a-11eb-96c4-b49948244cf1.png)

And TE with the workaround method requiring the quirk cited above:

![image](https://user-images.githubusercontent.com/16388068/104044582-bcd7ba00-519a-11eb-9ef5-5c9b97c56af6.png)

Both use a similar battery-less method, but with different branded chips as Wonder Boy in Monster World, here's my cart of that one for reference:

![image](https://user-images.githubusercontent.com/16388068/104045636-1b516800-519c-11eb-8760-6fdc2b7a5017.png)

**Conclusion of my testing**

So the current SRAM quirk method ostensibly works for other SRAM quirk games, but doesn't work for NBA Jam or NBA Jam TE. The SRAM is overwritten after title screen load if OSD is opened with either fresh data or the same data that isn't being read back correctly. Attached is a sav file after the initial write to SRAM occurred, and before the SRAM was overwritten the second time, so this can speed up testing for someone else to finish looking at it and fix it.

[NBA Jam (USA, Europe) (Rev 1).zip](https://github.com/MiSTer-devel/Genesis_MiSTer/files/5788615/NBA.Jam.USA.Europe.Rev.1.zip)

**Comment by birdybro** [comment]
I just added the SRAM_Quirk for NBA Jam Rev 1 and am compiling a new build with the change to test if the behavior at least for this one stays the same, as Rev 1 and TE have a slightly different layout on their PCB. It's possible things will only have to be adjusted for TE. Testing now, will push to my fork and submit a PR if this fixes it at least for Rev 1

**Comment by birdybro** [comment]
Given the way this has to be tested to be sure the potential changes are working, this took awhile:

sram_quirk for NBA Jam Rev1 test --> **Result** = Doesn't even drop a save file in the folder, so NBA Jam Rev1 is not an sram_quirk at all, and changing this of course would make things worse. Testing sram_quirk for NBA Jam T.E. to see whether or not I should change anything with that to fix it in this game... T.E. saves at title screen just like Rev1, it has the quirk originally anyways, so this is expected. Just want to see if it actually saves or not like claimed since I didn't test T.E. originally since I didn't know they were handled differently. T.E. also doesn't load the save and title screen overwrites save again (which is an empty save since it didn't read anything before).

Switching to eeprom_quirk for both NBA Jam Rev1 and NBA Jam T.E. test --> **Result** = Given that sram_quirk doesn't work for either games, I tested changing both to eeprom_quirk. neither saves.

Possibly related to this https://gendev.spritesmind.net/forum/viewtopic.php?t=206

```
all of them use 0x200000 and 0x200001 addresses for SCL and SDA lines
```

and this line:

https://github.com/MiSTer-devel/Genesis_MiSTer/blob/6cdd10376ec6c4a88d55d1571685307dfd908bd8/rtl/system.sv#L921

I'll try and figure out how to do a range of h200000 thru h200001 to see if that works...

**Comment by birdybro** [comment]
Followed up more on this. On the current release, with NBA Jam Tournament Edition no-intro matching AND NBA Jam Tournament Edition with the Everdrive SRAM hack, it still doesn't work either way. This is now out of my depth and what I was trying to do wasn't working. Rev1 saves through a different method given the PCB above most likely. Strange.

**Comment by Alex-Catalin** [comment]
I haven't tested with NBA JAM TE, only with the original NBA JAM game (all its revisions). This is indeed really strange, since the same rom files will produce good results when it comes to saving records on my Mega SD flashcart.

**Comment by birdybro** [comment]
Well the Mega SD and the Mega Everdrives are both running on original hardware. The reason this isn't working is because the behavior of the original hardware regarding how these roms handled saves is not implemented in Genesis_MiSTer core yet.

**Comment by Alex-Catalin** [comment]
This is one reason why I still keep my old NBA JAM copy :), which still has records saved on it even after more than 20 years now. It's one of my absolute all time favorite games and I can even say that I still enjoy the Mega Drive version more than even the new 2010 version, which is a very good game. I forced myself into liking the SNES variant of NBA JAM, but just couldn't, no music during a game like this? Really?

Anyway, thanks a lot for your effort and for your time on this!

**Comment by ekeeke** [comment]
Both carts use EEPROM chips for saving (X24C02 and 24C04), not "baterry-less" SRAM. The i2c protocol used to access these EEPROM chips is a little bit different from the one used for 24C01 EEPROM chip that can be found in Wonderboy Monster World or Rockman so it's possible Mister only implements the simplest protocol (or does not implement any i2c EEPROM and just implement some hack to prevent some eeprom games freezing on startip).
College Slam  uses even larger EEPROM which requires a third (extended) type of protocol. Additionaly to that, each company who have been using EEPROM in their cartridges for game saves (only Sega, Capcom, EA and Acclaim afaik) used different address/data lines to implement i2c protocol (which is based on one clock signal and one data in/out signal) so you need to implement different "mappers". Mega SD flashcart afaik implements all the known mappers.

Here is some doc I wrote years ago which describe the i2c protocols and various mappers: http://krikzz.com/pub/support/everdrive-md/v2/gen_eeprom.pdf

Otherwise, you can also use Genesis Plus GX sourcecode as reference for a more up-to-date list: https://github.com/ekeeke/Genesis-Plus-GX/blob/master/core/cart_hw/eeprom_i2c.c

**Comment by birdybro** [comment]
Thank you @ekeeke ! 

So I think then that the problem is probably that we currently have a "catch-all" "sram_quirk" (https://github.com/MiSTer-devel/Genesis_MiSTer/blob/6cdd10376ec6c4a88d55d1571685307dfd908bd8/Genesis.sv#L853) and this is insufficient if the methods are different for most of these games. This is helpful to someone I'm sure, I wish I was competent enough at coding to figure it out. I'm struggling to get to that point currently.

The MiSTer Genesis core currently handles these situations in only one of two ways, either it treats it as an "eeprom_quirk" (https://github.com/MiSTer-devel/Genesis_MiSTer/blob/6cdd10376ec6c4a88d55d1571685307dfd908bd8/rtl/system.sv#L917) or it treats it as an "sram_quirk"(https://github.com/MiSTer-devel/Genesis_MiSTer/blob/6cdd10376ec6c4a88d55d1571685307dfd908bd8/rtl/system.sv#L921). I also found this odd, since to me the chips used in the pictures above are eeprom chips, not sram. @Kitrinx also came up with an EEPROM support specifically for Pier Solar (https://github.com/MiSTer-devel/Genesis_MiSTer/blob/master/rtl/EEPROM_STM95.sv) but this is already handled by pier_quirk (https://github.com/MiSTer-devel/Genesis_MiSTer/blob/6cdd10376ec6c4a88d55d1571685307dfd908bd8/Genesis.sv#L868) in the core.

compare this with what ekeeke said above and linked:

https://github.com/ekeeke/Genesis-Plus-GX/blob/58758aef3d8e12189899d67344db3b95c6379561/core/cart_hw/eeprom_i2c.c#L122

You can see that there are far more exceptions and they are specific to the techniques used at different companies, and none of these exceptions are making the mistake of assuming it's "sram". It looks like solely plutiedev was used for how the eeprom saves should work in terms of starting register.

Thanks again for the help!

**Comment by ekeeke** [comment]
Looking at system_sv, it seems to me these 'quirks' (well at least sram_quirk and eeprom_quirk) are only hacks to get eeprom games booting fine in Mister but they don't allow these games to save correctly as there does not seem to be any access to i2c eeprom chip implemented (and I couldn't found any implementation of eeprom chip in the repository, besides eeprom chip used by pier solar, which unfortunately is a different type and use a different protocol ).

So there does not seem to be any 'easy' fix to your issue besides properly implementing i2c eeprom hardware and cartridge mappers in fpga.

**Comment by birdybro** [comment]
Yes, I believe there will have to be an equivalent method like the Pier Solar eeprom exception implemented to handle this. Definitely no small task and would be understandable if it wasn't a priority. :)
